### DISCUSSION TITLE
How to get current file URI

### DISCUSSION THREAD

**Discussion by msaquib17** [eclipse-theia/theia#11937] [Q&A]
How to get the URI(eg C:\\Users\win10\theia\pricing.msrule2) in theia? I know that there is a way using EditorManager using `editorManager.currentEditor.getResourceUri()` but that method is not working in my case because the currentEditor is getting the value undefined. Is there any alternative way to get the file Uri? Also let me know if there is no other way to get the file Uri.

**Comment by vince-fugnitto** [comment]
@msaquib17 given the assumptions that the file `pricing.msrule2` is opened and is an `EditorWidget` or derivation of one, the[`EditorManager`](https://eclipse-theia.github.io/theia/docs/next/index.html) should provide you the information you're looking for. Can you maybe describe your use-case in full? If you are still experiencing issues with `EditorManager` I would suggest to debug as well.

**Reply by msaquib17** [reply]
yes pricing.msrule2 is open (instance of EditorWidget) and I need its path. the currentEditor and activeEditor properties of EditorManager instance are becoming undefined and I'm looking for an alternative way to get the file path. Is there any way to get it?

**Reply by vince-fugnitto** [reply]
Are you able to share a code snippet of what you're attempting?

**Reply by msaquib17** [reply]
Sure,
```ts
import { InMemoryResources, Resource } from "@theia/core/lib/common";
import URI from "@theia/core/lib/common/uri";
import { SetExpressionTextRequestAction } from "../direct-expression-editor/set-expression-text-request-action";
import { FrontendApplication } from "@theia/core/lib/browser/frontend-application";
import { DiagramWidget } from "sprotty-theia";

@injectable()
export class MambaEditLabelUI extends EditLabelUI {

    protected inputElement: HTMLInputElement;
    protected types: string[] = ["String", "Int"];
    @inject(MonacoEditorFactory)
    protected monacoEditorFactory: MonacoEditorFactory;
    protected node: SModelElement;
    @inject(InMemoryResources)
    protected readonly inMemoryResources: InMemoryResources;
    protected resource: Resource;
    @inject(FrontendApplication) protected readonly application: FrontendApplication;

    constructor(@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher) {
        super();
        this.inputElement = document.createElement("input");
    }

    protected initializeContents(containerElement: HTMLElement): void {
        // do nothing
    }

    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    async show(root: Readonly<SModelRoot>, ...contextElementIds: string[]) {
        super.show(root, ...contextElementIds);
        if (this.isEditorEnabled()) {

            if (this.label) {
                if (this.application.shell.activeWidget instanceof DiagramWidget) {
                    const diagramWidget = this.application.shell.activeWidget;
                    const sourceUri = diagramWidget.uri.toString();

                    const resourceURI = "inmemory:/" + this.label.id + ".mexpression?sourceUri=" + sourceUri;
                    this.createResource(resourceURI, this.label.text);
                    const widget = await this.monacoEditorFactory.create(resourceURI);
                    if (widget instanceof MonacoEditorWidget) {
                        const dialog = new MonacoEditorWidgetDialog({ title: "Expression Editor" }, widget, this.onAccept.bind(this), this.onClose.bind(this));
                        dialog.open();
                    }
                }
            }
            } 
        }
```
Here in the function async show(), we have `const sourceUri = diagramWidget.uri.toString();`. Here diagramWidget is turning out to be undefined. I tried editorManager as well but its properties are also turning out to be undefined. I need to find some other way around.

**Reply by Yahud100** [reply]
if not solved u can try currentWidget instead of active widget in above code

**Comment by msujew** [comment]
@msaquib17 You could listen to the `ApplicationShell#onDidChangeCurrentWidget` event and always keep a reference to the last active `NavigatableWidget`. That way, you can always call `show` for the last widget which wasn't `undefined`.
